src/native/libs/System.Native/pal_io.c

```diff
@@ -1950,19 +1950,46 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
 }
 
 #if (HAVE_PREADV || HAVE_PWRITEV) && !defined(TARGET_WASM)
-static int GetAllowedVectorCount(int32_t vectorCount)
+static int GetAllowedVectorCount(IOVector* vectors, int32_t vectorCount)
 {
+#if defined(IOV_MAX)
+    const int IovMax = IOV_MAX;
+#else
+    // In theory all the platforms that we support define IOV_MAX,
+    // but we want to be extra safe and provde a fallback
+    // in case it turns out to not be true.
+    // 16 is low, but supported on every platform.
+    const int IovMax = 16;
+#endif
+
     int allowedCount = (int)vectorCount;
 
-#if defined(IOV_MAX)
-    if (IOV_MAX < allowedCount)
+    // We need to respect the limit of items that can be passed in iov.
+    // In case of writes, the managed code is responsible for handling incomplete writes.
+    // In case of reads, we simply returns the number of bytes read and it's up to the users.
+    if (IovMax < allowedCount)
+    {
+        allowedCount = IovMax;
+    }
+
+#if defined(TARGET_APPLE)
+    // For macOS preadv and pwritev can fail with EINVAL when the total length
+    // of all vectors overflows a 32-bit integer.
+    size_t totalLength = 0;
+    for (int i = 0; i < allowedCount; i++) 
     {
-        // We need to respect the limit of items that can be passed in iov.
-        // In case of writes, the managed code is reponsible for handling incomplete writes.
-        // In case of reads, we simply returns the number of bytes read and it's up to the users.
-        allowedCount = IOV_MAX;
+        totalLength += vectors[i].Count;
+
+        if (totalLength > INT_MAX && i > 0)
```

Comment:
Why is the `&& i > 0` necessary? Can `vectors[i].Count be greater than INT_MAX in practice?

Reply:
> Why is the `&& i > 0` necessary?

To avoid a situation where we pass `0` as `iovcnt` to the sys-call, it succeeds and the managed layer decides to continue the writes and ends up in an endless loop.

> Can `vectors[i].Count be greater than INT_MAX in practice?

Not right now, as `Memory.Length` is an Int32 as of today.

Reply:
> I'm missing something.

I expect that Count will be able to be more than Int32 one day. But most likely I am overthinking here.

Reply:
> I expect that Count will be able to be more than Int32 one day.

Ok. It'd be worth a comment, I think.

Reply:
I added an assert and removed the condition.